### ISSUE TITLE
[question] version 0.2.56 how to force LLAVA 1.5 CLIP use CUDA gpu?

### ISSUE THREAD

**Issue by eisneim** [abetlen/llama-cpp-python#1267]
version llama-cpp-python-0.2.56, how to enable CLIP offload to GPU? the llama part is fine, but CLIP is too slow
 my 3090 can do 50 token/s but total time would be tooo slow(92s), much slower than my Macbook M3 max(6s), 

i'v tried:  `CMAKE_ARGS="-DLLAMA_CUBLAS=on -DLLAVA_BUILD=on" pip install llama-cpp-python` but it does not work
```
Llama.generate: prefix-match hit

llama_print_timings:        load time =    5952.50 ms
llama_print_timings:      sample time =    2149.55 ms /  3438 runs   (    0.63 ms per token,  1599.41 tokens per second)
llama_print_timings: prompt eval time =       0.00 ms /     1 tokens (    0.00 ms per token,      inf tokens per second)
llama_print_timings:        eval time =   64415.00 ms /  3438 runs   (   18.74 ms per token,    53.37 tokens per second)
llama_print_timings:       total time =   90809.57 ms /  3439 tokens
-l=patio,furniture,chairs,umbrella,palm_trees,beach,ocean,relaxation,outdoor,sunny,tropical,vacation,leisure,tranquility,serenity,calm,peaceful,inviting,comfortable,picturesque,scenic,picturesque,beautiful,natural,idyllic,inviting,cozy,charming,inviti.jpg
cost: 91.75s {'prompt_tokens': 658, 'completion_tokens': 3438, 'total_tokens': 4096}
```

and here is my code
```python
chat_handler = Llava15ChatHandler(clip_model_path="/Users/teli/www/ml/vlm/llava_use/mmproj-model-f16.gguf")
llm = Llama(
  model_path="/Users/teli/www/ml/vlm/llava_use/llava-13b-q5_k.gguf",
  chat_handler=chat_handler,
  n_ctx=4096, # n_ctx should be increased to accomodate the image embedding
  logits_all=True,# needed to make llava work
  n_gpu_layers=-1, # 41
  temperature=0.1
)
prompt = "what is in the image? "
res = llm.create_chat_completion(
    messages = [
      {"role": "system", "content": "You are an assistant who perfectly describes images."},
      {
        "role": "user",
        "content": [
            {"type": "image_url", 
                "image_url": {"url": uri }
            },
            {"type" : "text", "text": prompt }
        ]
      }
    ]
  )
```

**Comment by ShuYiNiu** [comment]
I wonder that how you fixed the problem.

**Comment by eisneim** [comment]
the GPU is fully used the issue is that the generation keeps going until the context is full, so  just set max_tokens=64 fixed my problem, but when using Macbook M3 i don't need to do that and when the answer is short it will end it self, but same prompt on CUDA backend it will keep going if i didn't set max_tokens

**Comment by ShuYiNiu** [comment]
感谢回复！但我目前遇到的问题是total time并不等于前面几个(eval time...)的总和，不确定中间的时间损耗在哪里。我翻了很多issue后发现你的案例里也有这个问题，想问下你目前的total time能匹配上了吗？





> GPU 已完全使用，问题是生成一直持续到上下文已满，所以只需设置 max_tokens=64 即可解决我的问题
> 
> 但是当使用Macbook M3时我不必这样做，当答案很短时它会自行结束，但CUDA后端上的相同提示如果我没有设置max_tokens它会继续下去

**Comment by ShuYiNiu** [comment]
> the GPU is fully used the issue is that the generation keeps going until the context is full, so just set max_tokens=64 fixed my problem
> 
> but when using Macbook M3 i don't have to do that, and when the answer is short it will end itself, but same prompt on CUDA backend it will keep going if i didn't set max_tokens

Thank you for your reply! However, the issue I'm currently facing is that the total time does not equal the sum of the individual times (eval time, etc.) mentioned earlier. I'm unsure where the time loss occurs in between. After looking through many issues, I noticed this problem also exists in your example. I wanted to ask if your current total time is now matching the sum of the individual times?

**Comment by eisneim** [comment]
@ShuYiNiu on RTX 3090 total time is always 2 seconds for 54 tokens, it's fast enough for my use case; here are some output:

```
llama_print_timings:        load time =   40924.32 ms
llama_print_timings:      sample time =      30.06 ms /    54 runs   (    0.56 ms per token,  1796.17 tokens per second)
llama_print_timings: prompt eval time =       0.00 ms /     1 tokens (    0.00 ms per token,      inf tokens per second)
llama_print_timings:        eval time =     923.71 ms /    54 runs   (   17.11 ms per token,    58.46 tokens per second)
llama_print_timings:       total time =    1118.36 ms /    55 tokens
i1=cards,stickers,books,children,cartoons,characters,greetings,gifts,decorations,display,shelf,storage,organization,arrangement,variety,colors,shapes,sizes,designs,patterns,themes,emotions.jpg
cost: 1.95s {'prompt_tokens': 644, 'completion_tokens': 54, 'total_tokens': 698}
[76135 / 84068] snapshots_vlog19/房车旅行｜云南不大，全是神话！_275_20240303_1643-53_777.jpg
Llama.generate: prefix-match hit

llama_print_timings:        load time =   40924.32 ms
llama_print_timings:      sample time =      29.90 ms /    54 runs   (    0.55 ms per token,  1806.02 tokens per second)
llama_print_timings: prompt eval time =       0.00 ms /     1 tokens (    0.00 ms per token,      inf tokens per second)
llama_print_timings:        eval time =     924.24 ms /    54 runs   (   17.12 ms per token,    58.43 tokens per second)
llama_print_timings:       total time =    1121.41 ms /    55 tokens
JU=a_man_is_kneeling_down_on_the_ground,possibly_in_a_ditch,next_to_a_road_the_man_is_wearing_a_backpack,and_there_is_a_handbag_nearby_the_scene_appears_to_be_set_in_a_rural_area,with_a_grassy_field.jpg
cost: 1.94s {'prompt_tokens': 644, 'completion_tokens': 54, 'total_tokens': 698}


```
